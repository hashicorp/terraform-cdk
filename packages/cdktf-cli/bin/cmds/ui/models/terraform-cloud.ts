/* eslint-disable @typescript-eslint/no-unused-vars */
import * as path from "path";
import * as os from "os";
import * as fs from "fs";
import {
  Terraform,
  TerraformPlan,
  TerraformOutput,
  AbstractTerraformPlan,
} from "./terraform";
import { TerraformJsonConfigBackendRemote } from "../terraform-json";
import * as TerraformCloudClient from "@skorfmann/terraform-cloud";
import archiver from "archiver";
import { WritableStreamBuffer } from "stream-buffers";
import { SynthesizedStack } from "../../helper/synth-stack";
import { logger } from "../../../../lib/logging";
export class TerraformCloudPlan
  extends AbstractTerraformPlan
  implements TerraformPlan
{
  constructor(
    public readonly planFile: string,
    public readonly plan: { [key: string]: any },
    public readonly url: string
  ) {
    super(planFile, plan.resourceChanges, plan.outputChanges);
  }
}

export interface TerraformCredentialsItem {
  token: string;
}

export interface TerraformCredentials {
  [hostname: string]: TerraformCredentialsItem;
}
export interface TerraformCredentialsFile {
  credentials: TerraformCredentials;
}

const zipDirectory = (source: string): Promise<Buffer | false> => {
  const archive = archiver("tar", { gzip: true });
  const stream = new WritableStreamBuffer();

  return new Promise((resolve, reject) => {
    archive
      .directory(source, false)
      .on("error", (err) => reject(err))
      .on("end", () => resolve(stream.getContents()))
      .pipe(stream);
    archive.finalize();
  });
};

const wait = (ms = 1000) => {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
};

function BeautifyErrors(name: string) {
  return (target: any, propertyKey: string, descriptor: PropertyDescriptor) => {
    const isMethod = descriptor && descriptor.value instanceof Function;
    if (!isMethod) return;

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const originalMethod = descriptor!.value;
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    descriptor!.value = async function (...args: any[]) {
      try {
        return await originalMethod.apply(this, args);
      } catch (e) {
        if (
          e.response &&
          e.response.status >= 400 &&
          e.response.status <= 599
        ) {
          const errors = e.response.data?.errors as
            | Record<string, unknown>[]
            | undefined;
          logger.error(`Error in ${name}: ${JSON.stringify(e)}`);
          if (errors) {
            throw new Error(
              `${name}: Request to Terraform Cloud failed with status ${
                e.response.status
              }: ${errors.map((e) => JSON.stringify(e)).join(", ")}`
            );
          }
        } else {
          logger.warn(`Error in ${name}: ${JSON.stringify(e)}`);
        }
        throw e;
      }
    };

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    Object.defineProperty(target, propertyKey, descriptor!);
  };
}
export class TerraformCloud implements Terraform {
  private readonly terraformConfigFilePath = path.join(
    os.homedir(),
    ".terraform.d",
    "credentials.tfrc.json"
  );
  private readonly token: string;
  private readonly hostname: string;
  private readonly workspaceName: string;
  private readonly organizationName: string;
  private readonly client: TerraformCloudClient.TerraformCloud;
  private readonly isSpeculative: boolean;
  private configurationVersionId?: string;
  public readonly workDir: string;
  public run?: TerraformCloudClient.Run;

  constructor(
    public readonly stack: SynthesizedStack,
    public readonly config: TerraformJsonConfigBackendRemote,
    isSpeculative = false
  ) {
    if (!config.workspaces.name)
      throw new Error("Please provide a workspace name for Terraform Cloud");
    if (!config.organization)
      throw new Error("Please provide an organization for Terraform Cloud");
    this.workDir = stack.workingDirectory;

    this.hostname = config.hostname || "app.terraform.io";
    this.workspaceName = config.workspaces.name;
    this.organizationName = config.organization;
    this.isSpeculative = isSpeculative;
    if (config.token) {
      this.token = config.token;
    } else if (process.env.TERRAFORM_CLOUD_TOKEN) {
      this.token = process.env.TERRAFORM_CLOUD_TOKEN;
    } else {
      if (!fs.existsSync(this.terraformConfigFilePath))
        throw new Error("Please provide token for Terraform Cloud");
      const configFile = JSON.parse(
        fs.readFileSync(this.terraformConfigFilePath, "utf8")
      ) as TerraformCredentialsFile;
      this.token = configFile.credentials[this.hostname].token;
    }

    this.client = new TerraformCloudClient.TerraformCloud(this.token);
  }

  @BeautifyErrors("IsRemoteWorkspace")
  public async isRemoteWorkspace(): Promise<boolean> {
    const workspace = await this.workspace();
    return workspace.attributes.executionMode !== "local";
  }

  @BeautifyErrors("Init")
  public async init(): Promise<void> {
    if (
      fs.existsSync(
        path.join(process.cwd(), `terraform.${this.stack.name}.tfstate`)
      )
    )
      throw new Error(
        'Found a "terraform.tfstate" file in your current working directory. Please migrate the state manually to Terraform Cloud and delete the file afterwards. https://cdk.tf/migrate-state'
      );
    const workspace = await this.workspace();
    const version = await this.client.ConfigurationVersion.create(
      workspace.id,
      {
        data: {
          type: "configuration-version",
          attributes: {
            autoQueueRuns: false,
            speculative: this.isSpeculative,
          },
        },
      }
    );

    this.configurationVersionId = version.id;

    this.removeLocalTerraformDirectory();

    const zipBuffer = await zipDirectory(this.workDir);
    if (!zipBuffer)
      throw new Error("Couldn't upload directory to Terraform Cloud");

    await this.client.ConfigurationVersion.upload(
      version.attributes.uploadUrl,
      zipBuffer
    );

    // we might get an HTTP 422 error if we don't wait for the processing and try to run too early. see #647
    await this.waitForConfigurationVersionToBeReady();
  }

  @BeautifyErrors("Plan")
  public async plan(destroy = false): Promise<TerraformPlan> {
    if (!this.configurationVersionId)
      throw new Error("Please create a ConfigurationVersion before planning");
    const workspace = await this.workspace();
    const workspaceUrl = `https://app.terraform.io/app/${this.organizationName}/workspaces/${this.workspaceName}`;

    if (
      workspace.attributes.locked &&
      workspace.relationships?.lockedBy?.data?.type === "users"
    ) {
      throw new Error(
        `Can not plan, the workspace ${this.organizationName}/${this.workspaceName} is locked by a user. You can find more information at ${workspaceUrl}`
      );
    }

    if (
      workspace.attributes.locked &&
      workspace.relationships?.lockedBy?.data?.type === "runs"
    ) {
      throw new Error(
        `Can not plan, the workspace ${this.organizationName}/${this.workspaceName} is locked by a previous run, please wait until it's done. You can find more information at ${workspaceUrl}`
      );
    }

    if (workspace.attributes.locked) {
      throw new Error(
        `Can not plan, the workspace ${this.organizationName}/${
          this.workspaceName
        } is locked for an unknown reason: ${JSON.stringify(
          workspace.relationships?.lockedBy
        )}. You can find more information at ${workspaceUrl}`
      );
    }

    let result = await this.client.Runs.create({
      data: {
        attributes: {
          isDestroy: destroy,
          message: "cdktf",
        },
        relationships: {
          configurationVersion: {
            data: {
              id: this.configurationVersionId,
              type: "configuration-versions",
            },
          },
          workspace: {
            data: {
              id: workspace.id,
              type: "workspaces",
            },
          },
        },
      },
    });

    const pendingStates = ["pending", "plan_queued", "planning"];
    while (pendingStates.includes(result.attributes.status)) {
      result = await this.client.Runs.show(result.id);
      await wait(1000);
    }

    const url = `https://app.terraform.io/app/${this.organizationName}/workspaces/${this.workspaceName}/runs/${result.id}`;
    if (result.attributes.status === "errored") {
      throw new Error(`Error planning the run, please take a look at ${url}`);
    }

    const plan = await this.client.Plans.jsonOutput(
      result.relationships.plan.data.id
    );
    this.run = result;
    return new TerraformCloudPlan(
      "terraform-cloud",
      plan as unknown as any,
      url
    );
  }

  @BeautifyErrors("Deploy")
  public async deploy(
    _planFile: string,
    stdout: (chunk: Buffer) => any
  ): Promise<void> {
    if (!this.run)
      throw new Error(
        "Please create a ConfigurationVersion / Plan before deploying"
      );

    const deployingStates = ["confirmed", "apply_queued", "applying"];
    const runId = this.run.id;
    await this.client.Runs.action("apply", runId);
    let result = await this.client.Runs.show(runId);

    async function update(client: TerraformCloudClient.TerraformCloud) {
      const res = await client.Runs.show(runId);

      // fetch logs and update UI in the background
      client.Applies.logs(result.relationships.apply.data.id).then(({ data }) =>
        stdout(Buffer.from(data, "utf8"))
      );
      return res;
    }

    while (deployingStates.includes(result.attributes.status)) {
      result = await update(this.client);
      await wait(1000);
    }

    switch (result.attributes.status) {
      case "applied":
        break;
      default:
        throw new Error(`error: ${result.attributes.status}`);
    }
  }

  @BeautifyErrors("Destroy")
  public async destroy(stdout: (chunk: Buffer) => any): Promise<void> {
    if (!this.run)
      throw new Error(
        "Please create a ConfigurationVersion / Plan before destroying"
      );

    const destroyingStates = ["confirmed", "apply_queued", "applying"];
    const runId = this.run.id;
    await this.client.Runs.action("apply", runId);
    let result = await this.client.Runs.show(runId);

    async function update(client: TerraformCloudClient.TerraformCloud) {
      const res = await client.Runs.show(runId);

      // fetch logs and update UI in the background
      client.Applies.logs(result.relationships.apply.data.id).then(({ data }) =>
        stdout(Buffer.from(data, "utf8"))
      );
      return res;
    }

    while (destroyingStates.includes(result.attributes.status)) {
      result = await update(this.client);
      await wait(1000);
    }

    switch (result.attributes.status) {
      case "applied":
        break;
      default:
        throw new Error(`error: ${result.attributes.status}`);
    }
  }

  @BeautifyErrors("Version")
  public async version(): Promise<string> {
    return (await this.workspace()).attributes.terraformVersion;
  }

  @BeautifyErrors("Output")
  public async output(): Promise<{ [key: string]: TerraformOutput }> {
    const stateVersion = await this.client.StateVersions.current(
      (
        await this.workspace()
      ).id,
      true
    );
    if (!stateVersion.included) return {};

    const outputs = stateVersion.included.reduce((acc, output) => {
      acc[output.attributes.name] = {
        sensitive: output.attributes.sensitive,
        type: output.attributes.type,
        value: output.attributes.value,
      };

      return acc;
    }, {} as { [key: string]: TerraformOutput });

    return outputs;
  }

  @BeautifyErrors("Workspace")
  private async workspace() {
    try {
      return await this.client.Workspaces.showByName(
        this.organizationName,
        this.workspaceName
      );
    } catch (e) {
      if (e.response?.status === 404) {
        // return a more descriptive error message as http response is not descriptive enough
        // will not be touched by BeautifyErrors decorator
        throw new Error(
          `TerraformCloud returned an HTTP 404 error. Please make sure the configured organization (${this.organizationName}) and workspace (${this.workspaceName}) exist and you have the correct access rights.`
        );
      }
      throw e;
    }
  }

  private async isConfigurationVersionReady(): Promise<boolean> {
    if (!this.configurationVersionId)
      throw new Error(
        "No ConfigurationVersionId known. Cannot wait for ConfigurationVersion to become ready"
      );
    const configVersion = await this.client.ConfigurationVersion.show(
      this.configurationVersionId
    );
    const { status } = configVersion.attributes;
    switch (status) {
      case "uploaded":
        return true;
      case "errored": {
        const { error, errorMessage } = configVersion.attributes;
        logger.error(
          `ConfigurationVersion in Terraform Cloud has status "${status}". Returned config version was ${JSON.stringify(
            configVersion
          )}`
        );
        throw new Error(
          `Terraform Cloud ConfigurationVersion ${
            this.configurationVersionId
          } has status "errored" ${JSON.stringify({ error, errorMessage })}`
        );
      }
      case "pending": // fallthrough
      default:
        logger.debug(
          `ConfigurationVersion in Terraform Cloud not considered ready with status "${status}".`
        );
        return false;
    }
  }

  private async waitForConfigurationVersionToBeReady(
    timeoutMs = 60_000
  ): Promise<void> {
    logger.debug(
      "waiting for Configuration Version to be ready in Terraform Cloud"
    );
    let timeoutReached = false;
    const ready = async () => {
      while (!((await this.isConfigurationVersionReady()) && !timeoutReached)) {
        await wait(1000);
      }
    };
    const timeout = async () => {
      await wait(timeoutMs);
      timeoutReached = true;
      throw new Error(
        `Waiting for Terraform Cloud ConfigurationVersion to become ready timed out (timeout was ${timeoutMs}ms)`
      );
    };
    await Promise.race([ready(), timeout()]);
    logger.debug("Configuration Version is ready in Terraform Cloud");
  }

  private removeLocalTerraformDirectory() {
    try {
      fs.rmdirSync(
        path.resolve(this.stack.synthesizedStackPath, ".terraform"),
        { recursive: true }
      );
    } catch (error) {
      logger.debug(`Could not remove .terraform folder`, error);
    }
  }
}
