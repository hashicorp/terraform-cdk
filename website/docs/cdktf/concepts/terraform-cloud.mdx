---
page_title: Terraform Cloud
description: >-
  Use Terraform Cloud to keep your infrastructure state, apply policies, and execute infrastructure changes.
---

# Terraform Cloud

Terraform Cloud is HashiCorpâ€™s managed service offering. It eliminates the need for unnecessary tooling and documentation for practitioners, teams, and organizations to use Terraform in production.

Provision infrastructure in a remote environment that is optimized for the Terraform workflow.

You can use [these learn guides to learn more about Terraform Cloud](https://learn.hashicorp.com/collections/terraform/cloud-get-started).

## Why Terraform Cloud & Terraform Enterprise?

CDK for Terraform supports [Terraform Cloud](https://cloud.hashicorp.com/products/terraform), the managed service offering, and [Terraform Enterprise](/enterprise), the self-hosted distribution of Terraform Cloud.

Both Terraform Cloud and Terraform Enterprise can be used with CDKTF to [plan and execute infrastructure builds](/cloud-docs/run/remote-operations), [enforce organization-wide policies](/cloud-docs/sentinel), [estimate costs](/cloud-docs/cost-estimation), and [manage variables](/cloud-docs/workspaces/variables).

You may want to use Terraform Cloud when you need either of those functionalities or if you want a simple and reliable way to store the Terraform state. The default of writing the state into a local file means the risk of loosing your state by accident (the state contains sensitive information and should not be kept in version control), so we recommend Terraform Cloud as a way to persist the data.

## Set up CDKTF with Terraform Cloud

To set CDKTF up with Terraform Cloud you need to first create a [Terraform Cloud Account](https://app.terraform.io/public/signup/account) and [create a workspace](/cloud-docs/workspaces/creating).
As a workspace type you can either choose the CLI-driven workflow or the Version Control workflow. If you want to use the Version Control workflow you can find more information on [how to set it up](/cdktf/create-and-deploy/deployment-patterns#using-terraform-cloud) and [how a workflow might look like](/cdktf/create-and-deploy/deployment-patterns#deploy-with-terraform-and-terraform-cloud) in our deployment pattern docs.

Once you created the Terrafom Cloud workspace you can add a `RemoteBackend` to your CDKTF program:

```ts
import { Construct } from "constructs";
import { App, TerraformStack, RemoteBackend } from "cdktf";

class MyStack extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    // Remote Backend - /docs/backends/types/remote.html
    new RemoteBackend(this, {
      // You can omit this if you are not using a self-hosted Terraform enterprise instance
      hostname: "app.terraform.io",
      organization: "company",

      workspaces: {
        name: "my-app-prod",
      },
    });

    // define resources here
  }
}

const app = new App();
const myStack = new MyStack(app, "my-stack");
// Instead of defining the backend within the stack you can do it also on the outside like this:
// new RemoteBackend(myStack, {
//     hostname: "app.terraform.io",
//     organization: "company",

//     workspaces: {
//     name: "my-app-prod",
//     },
// });
// You just need to make sure you configure this before the app.synth() call
app.synth();
```

With this in place you are able to run your CDKTF commands against Terraform Cloud and Terraform Enterprise.

This all can be done automatically for Terraform Cloud by running `cdktf init` without passing the `--local` flag. CDKTF will use your stored [Terraform Cloud](/cloud) credentials to create a new workspace. If you have no stored Terraform Cloud credentials, CDKTF will ask you to login.

After you login, CDKTF creates a new scaffolded project that is configured to use Terraform state with a [remote backend](/language/settings/backends/remote). Where the Terraform operations will happen depends on the configuration of the Terraform Cloud Workspace settings. When you create the workspace as part of the `cdktf init` command, Terraform operations will be run locally by default. You can configure the Terraform Cloud workspace to use remote operations instead (details below).

## Managing Variables and Secrets through Terraform Cloud

You can use [Variables](/cloud-docs/workspaces/variables) to define variables and secrets to use during the execution of a plan / apply / destroy operation in Terraform Cloud. You can manage multiple secrets across multiple workspaces by using [variable sets](/cloud-docs/workspaces/variables/managing-variables#variable-sets).

To be able to use variables you have to enable [remote operations](/cloud-docs/run/remote-operations) for your Terraform Cloud workspace, this means that not a local Terraform CLI but Terraform Cloud is used to execute the run.

To declare variables you can use the `TerraformVariable` construct which renders an [Input Variable](/language/values/variables):

```ts
import { Construct } from "constructs";
import { App, TerraformStack, TerraformVariable } from "cdktf";

class MyStack extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    // You can configure the variable 'my-var' in Terraform Cloud and use the value throughout your CDKTF application
    const value = new TerraformVariable(this, "my-var", {
      default: "The default value to use",
      description: "",
      nullable: false, // if passing no
      sensitive: true,
      type: "string",
    });

    // define resources here
  }
}
```

Please bear in mind that the return value is an object with various token representations of the value as attributes. You can pass it as a string via `value.stringValue`, as a number via `value.numberValue`, or as a list via `value.listValue`. This means during synthesization there is no real value in the field, only a [Token](/cdktf/concepts/tokens). Tokens can not be used in dynamic checks during the runtime, e.g. `if (value.listValue.length > 42) {` will always return `false` as tokenized lists have a static length of one item.

## Continuous Integration

To run Terraform Cloud in a CI workflow you can either use [Terraform Clouds VCS option directly](/cdktf/create-and-deploy/deployment-patterns#deploy-with-terraform-and-terraform-cloud) or use a [general-purpose CI to trigger the run in Terraform Cloud](/cdktf/create-and-deploy/deployment-patterns#github-actions-ci-and-terraform-cloud). To learn more about when to use what visit our [Deployment Pattern docs](/cdktf/create-and-deploy/deployment-patterns).

## Policy Enforcement

Terraform Cloud lets you define [Sentinel Policies](/cloud-docs/sentinel) that can be applied across one or multiple Terraform Cloud workspaces to ensure the provisioned infrastructure follows company-wide security standards / best practices. You can use them with CDKTF if Terraform Cloud is configured as a backend.

You can use [Run Tasks](/cloud-docs/workspaces/settings/run-tasks) to integrate any pre-apply-checks or post-apply-notifications to your CI workflow via a webhook interface.
