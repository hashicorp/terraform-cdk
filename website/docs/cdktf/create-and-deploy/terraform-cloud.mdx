---
page_title: Connect to Terraform Cloud / Enterprise - CDK for Terraform
description: >-
  Use Terraform Cloud with CDKTF to store Terraform state remotely, manage secrets, apply policies, and more.
---

# Connect to Terraform Cloud / Enterprise

This page explains how to connect Terraform Cloud / Enterprise to your CDK for Terraform (CDKTF) application and the benefits of using these products together.

> **Hands On:** Try our [Get Started - Terraform Cloud](/terraform/tutorials/cloud-get-started) tutorials.

## What are Terraform Cloud and Terraform Enterprise?

CDKTF supports [Terraform Cloud](https://cloud.hashicorp.com/products/terraform) and [Terraform Enterprise](/terraform/enterprise).

Terraform Cloud is a SaaS application that runs Terraform in a stable, remote environment and securely stores state and secrets. It includes a user interface that helps you better understand your Terraform operations and resources, allows you to define role-based access controls, and offers a private registry for sharing [modules](/terraform/cdktf/concepts/modules) and [providers](/terraform/cdktf/concepts/providers). Terraform Cloud also integrates with the Terraform CLI and connects to common version control systems (VCS) like GitHub, GitLab, and Bitbucket. When you connect a Terraform Cloud workspace to a VCS repository, new commits and changes can automatically trigger Terraform plans. Terraform Cloud also offers an API, allowing you to integrate it into existing workflows.

Terraform Enterprise lets you set up a self-hosted distribution of Terraform Cloud and is ideal for organizations with strict security and compliance requirements.

## When to use Terraform Cloud or Terraform Enterprise?

Terraform uses persisted state data to keep track of the real-world resources it manages. By default, Terraform writes state to a local file. We recommend integrating with Terraform Cloud to store state remotely. This prevents accidental loss of locally stored state and lets multiple people access the state data so they can work together on that collection of infrastructure resources.

Other benefits of Terraform Cloud and Terraform Enterprise include the ability to run Terraform remotely, [manage variables and secrets](#managing-variables-and-secrets), and enforce [policies for sets of infrastructure](#policy-enforcement).

## Set up CDKTF with Terraform Cloud / Enterprise

After you sign up for a [Terraform Cloud Account](https://app.terraform.io/public/signup/account) / create an account in your Terraform Enterprise instance, you must connect your CDKTF project to one or more [workspaces](/terraform/cloud-docs/workspaces).

Workspaces function like working directories for distinct Terraform configurations and are associated with a Terraform configuration and its state file. They can also contain variables that you can use to manage credentials.

Workspaces can have one of three workflows that determine how Terraform Cloud / Enterprise interacts with your CDKTF application: CLI-driven, Version control, or API-driven (a more advanced option).

### CLI-driven Workflow

You can configure the Terraform Cloud [CLI-driven workflow](/terraform/cloud-docs/run/cli) for both new and existing CDKTF projects. This workflow lets you run CDKTF CLI commands to deploy your application and run Terraform operations remotely on Terraform Cloud / Enterprise.

#### Connect New CDKTF Projects

First use `cdktf login` to log into Terraform Cloud or your Terraform Enterprise instance:

<Tabs>
  <Tab heading="Terraform Cloud" group="tfc">

    This command will take you through an interactive session to log in.

    ```shell-session
    $ cdktf login
    ```

    If you already have a Terraform Cloud API token, you can pass it into stdin:

    ```shell-session
    cat my-token.txt | cdktf login
    ```

  </Tab>

  <Tab heading="Terraform Enterprise" group="tfe">

    This command will take you through an interactive session to log in.

    ```shell-session
    $ cdktf login --tfe-hostname tfe.my-company.com
    ```

    If you already have a Terraform Enterprise API token, you can pass it into stdin:

    ```shell-session
    cat my-token.txt | cdktf login --tfe-hostname tfe.my-company.com
    ```

  </Tab>
</Tabs>

1. Run `cdktf init` without passing the `--local` flag. This creates a new template CDKTF project for your chosen programming language. It will be the basis for a new application with the necessary file structure for you to start defining infrastructure. If you're using Terraform Enterprise, you must also pass the `--tfe-hostname` flag with the hostname of your enterprise instance.
2. Choose a project language and provide a project name and description.
3. Select a Terraform Cloud organization and a name for your workspace. CDKTF will create a [CloudBackend](/terraform/cli/cloud/settings#the-cloud-block) in your project.
4. You can choose whether to send crash reports to the CDKTF team and select the providers you want to use.

CDKTF creates both a scaffolded project in your chosen language and a new Terraform Cloud workspace on first apply. Your project is connected to the workspace and you can begin using CDKTF and Terraform Cloud together.
By default, the workspace is set to Remote execution mode, which means Terraform runs remotely on Terraform Cloud / Enterprise. If you want Terraform to run locally, you need to set the workspace to [Local Execution mode](/terraform/cloud-docs/workspaces/settings#execution-mode).

#### Connect Existing CDKTF Projects

For existing projects, do the following:

1. Create a Terraform Cloud workspace with the CLI-driven workflow.
2. Add the `RemoteBackend` to your CDKTF application. The following example connects the application to a Terraform Cloud / Enterprise workspace called `my-app-prod`.

   ```typescript
   import { Construct } from "constructs";
   import { App, TerraformStack, RemoteBackend } from "cdktf";

   class MyStack extends TerraformStack {
     constructor(scope: Construct, name: string) {
       super(scope, name);
       // Remote Backend - /docs/backends/types/remote.html
       new RemoteBackend(this, {
         // Only required for self-hosted Terraform Enterprise instances.
         // Defaults to "app.terraform.io".
         hostname: "app.terraform.io",
         organization: "company",

         workspaces: {
           name: "my-app-prod",
         },
       });

       // define resources here
     }
   }

   const app = new App();
   const myStack = new MyStack(app, "my-stack");
   // You can also define the backend outside of the stack.
   // new RemoteBackend(myStack, {
   //     hostname: "app.terraform.io",
   //     organization: "company",

   //     workspaces: {
   //     name: "my-app-prod",
   //     },
   // });
   // You must configure the RemoteBackend before the app.synth() call
   app.synth();
   ```

3. Run `cdktf login` / `cdktf login --tfe-hostname=tfe.my-company.com` and log in to Terraform Cloud / Enterprise.

Your CDKTF application is connected to the Terraform Cloud / Enterprise workspace you specified in the `RemoteBackend`. You must set the workspace to [Remote Execution mode](/terraform/cloud-docs/workspaces/settings#execution-mode) if you want to store state and run Terraform operations remotely.

### VCS-driven Workflow

Check your synthesized code into version control and connect your Terraform Cloud / Enterprise workspace to that repository with the VCS-driven workflow. You can configure the workspace to trigger Terraform runs based on merges and commits to the repository. Refer to [Deployment Patterns](/terraform/cdktf/create-and-deploy/deployment-patterns#deploy-with-terraform-and-terraform-cloud) for more details about how to use CDKTF with the VCS-driven workflow.

## Managing Variables and Secrets

Terraform Cloud / Enterprise [variables](/terraform/cloud-docs/workspaces/variables) let you define the variables and secrets that Terraform uses during remote operations. You can set variables specifically for each workspace or you can create variable sets to reuse the same variables across multiple workspaces. For example, you could define a variable set of provider credentials and automatically apply it to all of the workspaces using that provider.

To use variables, set your workspace to [Remote Execution Mode](/terraform/cloud-docs/workspaces/settings#execution-mode). Terraform can only access workspace variables when executing remotely on Terraform Cloud / Enterprise.

2. Use the `TerraformVariable` construct to declare variables in your CDKTF application. This creates an undefined [Terraform input variable](/terraform/language/values/variables). The following example demonstrates how to create a `TerraformVariable` called `my-var`.

   ```typescript
   import { Construct } from "constructs";
   import { App, TerraformStack, TerraformVariable } from "cdktf";

   class MyStack extends TerraformStack {
     constructor(scope: Construct, name: string) {
       super(scope, name);
       // You can define 'my-var' in Terraform Cloud and use the value in your application.
       const value = new TerraformVariable(this, "my-var", {
         default: "The default value to use",
         description: "",
         nullable: false, // if passing no
         sensitive: true,
         type: "string",
       });

       // define resources here
     }
   }
   ```

3. Add variables that you declared in your CDKTF application to your Terraform Cloud / Enterprise workspace. Refer to [Add a Variable](/terraform/cloud-docs/workspaces/variables/managing-variables#add-a-variable) for details. This lets Terraform access the value during operations.

### Variable Return Values

The return value for `TerraformVariable` is an object with various representations of the value as attributes. You can pass this object as a string with `value.stringValue`, as a number with `value.numberValue`, or as a list with `value.listValue`.

This means that there is no actual value in the variable field when CDKTF synthesizes your application. Instead, CDKTF uses a [Token](/terraform/cdktf/concepts/tokens), which represents a value that is unknown until Terraform applies your configuration. You cannot use tokens in dynamic checks during runtime. For example, `if (value.listValue.length > 42) {` always returns `false` because tokenized lists have a static length of one item.

## Continuous Integration

To run Terraform Cloud / Enterprise in a CI workflow, you can either use [Terraform Cloud / Enterprise's VCS-driven workflow](/terraform/cdktf/create-and-deploy/deployment-patterns#deploy-with-terraform-and-terraform-cloud) or use a [general-purpose CI to trigger the run in Terraform Cloud / Enterprise](/terraform/cdktf/create-and-deploy/deployment-patterns#github-actions-ci-and-terraform-cloud).

## Policy Enforcement

+-> **Note:** Refer to [Terraform Cloud pricing](https://www.hashicorp.com/products/terraform/pricing) for more information about Sentinel policies and run tasks.

You can define [Sentinel policies](/terraform/cloud-docs/policy-enforcement) for one or multiple Terraform Cloud workspaces to ensure that your infrastructure follows company-wide security standards and best practices. You can use Sentinel policies with CDKTF if Terraform Cloud is configured as a [Remote Backend](/terraform/cdktf/concepts/remote-backends).

You can also use [Run Tasks](/terraform/cloud-docs/workspaces/settings/run-tasks) to directly integrate third-party tools and services at certain stages in the Terraform Cloud run lifecycle. Terraform Cloud uses the status response from the third-party tool to determine if a run should proceed.
