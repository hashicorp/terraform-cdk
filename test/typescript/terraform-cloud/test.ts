// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { TestDriver, onPosix } from "../../test-helper";
import { TerraformCloud } from "@skorfmann/terraform-cloud";
import * as crypto from "crypto";
import { readFileSync } from "fs-extra";

const { TERRAFORM_CLOUD_TOKEN, GITHUB_RUN_NUMBER, TERRAFORM_VERSION } =
  process.env;
const withAuth = TERRAFORM_CLOUD_TOKEN ? onPosix : it.skip;

if (!TERRAFORM_CLOUD_TOKEN) {
  console.log("TERRAFORM_CLOUD_TOKEN is undefined, skipping authed tests");
}

process.on("uncaughtException", (err) => {
  console.dir(err);
  console.log("UNHANDLED");
  process.exit(1);
});

process.on("unhandledRejection", (err) => {
  console.dir(err);
  console.log("UNHANDLED");
  throw err;
});

// Below tests are disabled on windows because they fail due to networking issues
describe("full integration test", () => {
  let driver: TestDriver;
  let workspaceName: string;
  const orgName = "cdktf";

  beforeEach(async () => {
    workspaceName = `${GITHUB_RUN_NUMBER}-${crypto
      .randomBytes(10)
      .toString("hex")}`;
    driver = new TestDriver(__dirname, {
      TERRAFORM_CLOUD_WORKSPACE_NAME: workspaceName,
      TERRAFORM_CLOUD_ORGANIZATION: orgName,
    });
    await driver.setupTypescriptProject();
    driver.copyFolders("fixtures");
    console.log(driver.workingDirectory);
  });

  withAuth("deploy in Terraform Cloud", async () => {
    const client = new TerraformCloud(TERRAFORM_CLOUD_TOKEN);

    await client.Workspaces.create(orgName, {
      data: {
        attributes: {
          name: workspaceName,
          executionMode: "remote",
          terraformVersion: TERRAFORM_VERSION,
        },
        type: "workspaces",
      },
    });

    expect(await driver.deploy(["source-stack"])).toContain("Apply complete!");
    await client.Workspaces.deleteByName(orgName, workspaceName);
  });

  withAuth("deploy locally and then in Terraform Cloud", async () => {
    const client = new TerraformCloud(TERRAFORM_CLOUD_TOKEN);

    await client.Workspaces.create(orgName, {
      data: {
        attributes: {
          name: workspaceName,
          executionMode: "remote",
          terraformVersion: TERRAFORM_VERSION,
        },
        type: "workspaces",
      },
    });

    process.env.TF_EXECUTE_LOCAL = "true";
    await driver.deploy(["source-stack"], "before-migration.json");
    process.env.TF_EXECUTE_LOCAL = undefined;
    await driver.deploy(["source-stack"], "after-migration.json", [
      "--migrate-state",
    ]);

    expect(readFileSync("before-migration.json")).toEqual(
      readFileSync("after-migration.json")
    );

    await client.Workspaces.deleteByName(orgName, workspaceName);
  });

  // Only the origin stack is in TFC, the consumer stack is local
  withAuth(
    "deploy with cross stack reference origin in Terraform Cloud",
    async () => {
      const client = new TerraformCloud(TERRAFORM_CLOUD_TOKEN);

      await client.Workspaces.create(orgName, {
        data: {
          attributes: {
            name: workspaceName,
            executionMode: "remote",
            terraformVersion: TERRAFORM_VERSION,
          },
          type: "workspaces",
        },
      });

      await driver.deploy(["source-stack", "consumer-stack"]);
      driver.output("source-stack", "outputs.tmp.json", true);
      const outputs = JSON.parse(readFileSync("outputs.tmp.json").toString());

      await client.Workspaces.deleteByName(orgName, workspaceName);

      expect(driver.readLocalFile("consumer-file.txt")).toEqual(
        outputs["source-stack"].password_output
      );
    }
  );
});
