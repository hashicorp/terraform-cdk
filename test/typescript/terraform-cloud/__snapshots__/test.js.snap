// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`full integration test deploy in Terraform Cloud 1`] = `
"Deploying Stack: hello-terra
Resources
 ✔ NULL_RESOURCE       test                null_resource.test


Summary: 1 created, 0 updated, 0 destroyed.

Output: isAssetPresent = yes
        output = constant value
"
`;
